File: sdk/python/coreauth/_http.py
"""Internal HTTP transport for CoreAuth SDK."""

from __future__ import annotations

from typing import Any, Optional

import httpx

from .exceptions import (
    ApiError,
    AuthenticationError,
    ConflictError,
    CoreAuthError,
    ForbiddenError,
    NotFoundError,
    RateLimitError,
    ValidationError,
)


class HttpClient:
    """Low-level HTTP client wrapping httpx."""

    def __init__(self, base_url: str, token: Optional[str] = None) -> None:
        self._base_url = base_url.rstrip("/")
        self._token: Optional[str] = token
        self._client = httpx.Client(timeout=30.0)

    def set_token(self, token: str) -> None:
        self._token = token

    def clear_token(self) -> None:
        self._token = None

    def _headers(self) -> dict[str, str]:
        headers: dict[str, str] = {"Content-Type": "application/json"}
        if self._token:
            headers["Authorization"] = f"Bearer {self._token}"
        return headers

    def _form_headers(self) -> dict[str, str]:
        headers: dict[str, str] = {"Content-Type": "application/x-www-form-urlencoded"}
        if self._token:
            headers["Authorization"] = f"Bearer {self._token}"
        return headers

    def _handle_response(self, resp: httpx.Response) -> Any:
        if resp.status_code == 204:
            return None
        if 200 <= resp.status_code < 300:
            if not resp.content:
                return None
            return resp.json()
        # Error handling
        try:
            body = resp.json()
            error = body.get("error", "")
            message = body.get("message", "")
        except Exception:
            error = ""
            message = resp.text

        error_map = {
            400: ValidationError,
            401: AuthenticationError,
            403: ForbiddenError,
            404: NotFoundError,
            409: ConflictError,
            429: RateLimitError,
        }
        cls = error_map.get(resp.status_code, ApiError)
        if cls is ApiError:
            raise ApiError(resp.status_code, error, message)
        raise cls(error, message)

    def get(self, path: str, params: Optional[dict[str, Any]] = None) -> Any:
        url = f"{self._base_url}{path}"
        try:
            resp = self._client.get(url, headers=self._headers(), params=params)
        except httpx.HTTPError as e:
            raise CoreAuthError(f"Request failed: {e}") from e
        return self._handle_response(resp)

    def post(self, path: str, json: Optional[dict[str, Any]] = None) -> Any:
        url = f"{self._base_url}{path}"
        try:
            resp = self._client.post(url, headers=self._headers(), json=json)
        except httpx.HTTPError as e:
            raise CoreAuthError(f"Request failed: {e}") from e
        return self._handle_response(resp)

    def post_form(self, path: str, data: dict[str, Any]) -> Any:
        url = f"{self._base_url}{path}"
        try:
            resp = self._client.post(url, headers=self._form_headers(), data=data)
        except httpx.HTTPError as e:
            raise CoreAuthError(f"Request failed: {e}") from e
        return self._handle_response(resp)

    def put(self, path: str, json: Optional[dict[str, Any]] = None) -> Any:
        url = f"{self._base_url}{path}"
        try:
            resp = self._client.put(url, headers=self._headers(), json=json)
        except httpx.HTTPError as e:
            raise CoreAuthError(f"Request failed: {e}") from e
        return self._handle_response(resp)

    def patch(self, path: str, json: Optional[dict[str, Any]] = None) -> Any:
        url = f"{self._base_url}{path}"
        try:
            resp = self._client.patch(url, headers=self._headers(), json=json)
        except httpx.HTTPError as e:
            raise CoreAuthError(f"Request failed: {e}") from e
        return self._handle_response(resp)

    def delete(self, path: str, json: Optional[dict[str, Any]] = None) -> Any:
        url = f"{self._base_url}{path}"
        try:
            resp = self._client.delete(url, headers=self._headers(), json=json)
        except httpx.HTTPError as e:
            raise CoreAuthError(f"Request failed: {e}") from e
        return self._handle_response(resp)

    def close(self) -> None:
        self._client.close()

File: sdk/go/coreauth/http.go
package coreauth

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

type httpClient struct {
	baseURL    string
	token      string
	httpClient *http.Client
}

func newHTTPClient(baseURL string, hc *http.Client) *httpClient {
	return &httpClient{
		baseURL:    strings.TrimRight(baseURL, "/"),
		httpClient: hc,
	}
}

func (c *httpClient) setToken(token string) {
	c.token = token
}

func (c *httpClient) clearToken() {
	c.token = ""
}

func (c *httpClient) doRequest(ctx context.Context, method, path string, body io.Reader, contentType string) (json.RawMessage, error) {
	u := c.baseURL + path
	req, err := http.NewRequestWithContext(ctx, method, u, body)
	if err != nil {
		return nil, &CoreAuthError{Message: fmt.Sprintf("failed to create request: %v", err)}
	}
	if contentType != "" {
		req.Header.Set("Content-Type", contentType)
	}
	if c.token != "" {
		req.Header.Set("Authorization", "Bearer "+c.token)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, &CoreAuthError{Message: fmt.Sprintf("request failed: %v", err)}
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, &CoreAuthError{Message: fmt.Sprintf("failed to read response: %v", err)}
	}

	if resp.StatusCode == 204 {
		return nil, nil
	}
	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
		if len(respBody) == 0 {
			return nil, nil
		}
		return json.RawMessage(respBody), nil
	}

	// Parse error
	apiErr := &ApiError{StatusCode: resp.StatusCode}
	var errBody struct {
		Error   string `json:"error"`
		Message string `json:"message"`
	}
	if json.Unmarshal(respBody, &errBody) == nil {
		apiErr.ErrorCode = errBody.Error
		apiErr.Message = errBody.Message
	} else {
		apiErr.Message = string(respBody)
	}
	return nil, apiErr
}

func (c *httpClient) get(ctx context.Context, path string, params map[string]string) (json.RawMessage, error) {
	if len(params) > 0 {
		v := url.Values{}
		for k, val := range params {
			if val != "" {
				v.Set(k, val)
			}
		}
		if encoded := v.Encode(); encoded != "" {
			path = path + "?" + encoded
		}
	}
	return c.doRequest(ctx, http.MethodGet, path, nil, "application/json")
}

func (c *httpClient) post(ctx context.Context, path string, payload any) (json.RawMessage, error) {
	var body io.Reader
	if payload != nil {
		b, err := json.Marshal(payload)
		if err != nil {
			return nil, &CoreAuthError{Message: fmt.Sprintf("failed to marshal request: %v", err)}
		}
		body = bytes.NewReader(b)
	}
	return c.doRequest(ctx, http.MethodPost, path, body, "application/json")
}

func (c *httpClient) postForm(ctx context.Context, path string, data url.Values) (json.RawMessage, error) {
	return c.doRequest(ctx, http.MethodPost, path, strings.NewReader(data.Encode()), "application/x-www-form-urlencoded")
}

func (c *httpClient) put(ctx context.Context, path string, payload any) (json.RawMessage, error) {
	var body io.Reader
	if payload != nil {
		b, err := json.Marshal(payload)
		if err != nil {
			return nil, &CoreAuthError{Message: fmt.Sprintf("failed to marshal request: %v", err)}
		}
		body = bytes.NewReader(b)
	}
	return c.doRequest(ctx, http.MethodPut, path, body, "application/json")
}

func (c *httpClient) patch(ctx context.Context, path string, payload any) (json.RawMessage, error) {
	var body io.Reader
	if payload != nil {
		b, err := json.Marshal(payload)
		if err != nil {
			return nil, &CoreAuthError{Message: fmt.Sprintf("failed to marshal request: %v", err)}
		}
		body = bytes.NewReader(b)
	}
	return c.doRequest(ctx, http.MethodPatch, path, body, "application/json")
}

func (c *httpClient) del(ctx context.Context, path string, payload any) (json.RawMessage, error) {
	var body io.Reader
	if payload != nil {
		b, err := json.Marshal(payload)
		if err != nil {
			return nil, &CoreAuthError{Message: fmt.Sprintf("failed to marshal request: %v", err)}
		}
		body = bytes.NewReader(b)
	}
	return c.doRequest(ctx, http.MethodDelete, path, body, "application/json")
}
